### PR TITLE
Auto fire spit and spitter primordial.

### PR DESCRIPTION
**Pull request by hyper2snyper**
## About The Pull Request
Rewrites xeno spit to use the autofire component. 
Primordial Spitter gets super BRRRRRT.

Dont get used to it, I'm not becoming a xeno coder.

## Why It's Good For The Game

Carpal tunnel spit bad. Brrt spit good.

## Changelog
:cl:
add: Xeno spit now uses autofire.
add: Primordial Spitter: now with an incredible automatic firerate.
/:cl:

### CODE REVIEW

**Comment by Moggram** [comment]
>
>
>>Dont get used to it, I'm not becoming a xeno coder.

That's what all the DMs say....

File: code/_onclick/click.dm [hunk 1]
```diff
@@ -333,7 +333,7 @@ if(selected_ability.target_flags & flagname && !istype(A, typepath)){\
 		A = ability_target(A)
 		if(selected_ability.can_use_ability(A))
 			selected_ability.use_ability(A)
-			return TRUE
+			return CHECK_BITFIELD(selected_ability.use_state_flags, XACT_DO_AFTER_ATTACK) ? FALSE : TRUE
```
Comment (TiviPlus) [inline code/_onclick/click.dm:336]: suggestions are broken but you can jst use

return !(var & flag)

File: code/modules/mob/living/carbon/xenomorph/castes/spitter/castedatum_spitter.dm [hunk 2]
```diff
@@ -148,4 +148,30 @@
 	// *** Ranged Attack *** //
 	spit_delay = 0.5 SECONDS
 
+/datum/xeno_caste/spitter/primordial
+	upgrade_name = "Primordial"
+	caste_desc = "Master of ranged combat, this xeno knows no equal."
```
Comment (TiviPlus) [inline code/modules/mob/living/carbon/xenomorph/castes/spitter/castedatum_spitter.dm:153]: You need an primordial_upgrade_name and a hive_upgrade datum for this too

**Review by TiviPlus — CHANGES_REQUESTED** [review]
@BraveMole autofire code

File: code/modules/mob/living/carbon/xenomorph/abilities.dm [hunk 3]
```diff
@@ -816,17 +816,29 @@
 	mechanics_text = "Spit neurotoxin or acid at your target up to 7 tiles away."
 	ability_name = "xeno spit"
 	keybind_signal = COMSIG_XENOABILITY_XENO_SPIT
-	use_state_flags = XACT_USE_LYING|XACT_USE_BUCKLED
+	use_state_flags = XACT_USE_LYING|XACT_USE_BUCKLED|XACT_DO_AFTER_ATTACK
 	target_flags = XABB_MOB_TARGET
+	var/current_target
```
Comment (BraveMole) [inline code/modules/mob/living/carbon/xenomorph/abilities.dm:821]: autodoc

File: code/modules/mob/living/carbon/xenomorph/abilities.dm [hunk 4]
```diff
@@ -860,15 +876,30 @@
...
+	if(!xeno.check_state() || xeno.ammo?.spit_cost > xeno.plasma_stored)
 		return fail_activate()
+	set_target(get_turf_on_clickcatcher(object, xeno, params))
+	if(!current_target)
+		return
+	xeno.visible_message(span_xenowarning("\The [xeno] spits at \the [current_target]!"), \
+	span_xenowarning("We spit at \the [current_target]!") )
+
+	SEND_SIGNAL(src, COMSIG_XENO_FIRE)
+	xeno?.client?.mouse_pointer_icon = 'icons/effects/supplypod_target.dmi'
```
Comment (BraveMole) [inline code/modules/mob/living/carbon/xenomorph/abilities.dm:897]: another icon would be great

File: code/modules/mob/living/carbon/xenomorph/abilities.dm [hunk 5]
```diff
@@ -816,17 +816,29 @@
...
 	ability_name = "xeno spit"
 	keybind_signal = COMSIG_XENOABILITY_XENO_SPIT
-	use_state_flags = XACT_USE_LYING|XACT_USE_BUCKLED
+	use_state_flags = XACT_USE_LYING|XACT_USE_BUCKLED|XACT_DO_AFTER_ATTACK
 	target_flags = XABB_MOB_TARGET
+	var/current_target
+
+/datum/action/xeno_action/activable/xeno_spit/give_action(mob/living/L)
+	. = ..()
+	AddComponent(/datum/component/automatedfire/autofire, get_cooldown(), _fire_mode = GUN_FIREMODE_AUTOMATIC,  _callback_reset_fire = CALLBACK(src, .proc/reset_fire), _callback_fire = CALLBACK(src, .proc/fire))
```
Comment (BraveMole) [inline code/modules/mob/living/carbon/xenomorph/abilities.dm:825]: need a remove action as well

**Review by BraveMole — CHANGES_REQUESTED** [review]

File: code/modules/mob/living/carbon/xenomorph/abilities.dm [hunk 6]
```diff
@@ -824,6 +824,11 @@
 	. = ..()
 	AddComponent(/datum/component/automatedfire/autofire, get_cooldown(), _fire_mode = GUN_FIREMODE_AUTOMATIC,  _callback_reset_fire = CALLBACK(src, .proc/reset_fire), _callback_fire = CALLBACK(src, .proc/fire))
 
+/datum/action/xeno_action/activable/xeno_spit/remove_action(mob/living/L)
+	. = ..()
+	GetComponent(/datum/component/automatedfire/autofire).RemoveComponent()
+
```
Comment (TiviPlus) [inline code/modules/mob/living/carbon/xenomorph/abilities.dm:830]: qdel the component dont call removecomonent

**Comment by Lumipharon** [comment]
I'm guessing it's this PR, but scorpion minions don't seem to be able to shoot anymore... This is the only thing tested merged right now afaik.

**Review by BraveMole — CHANGES_REQUESTED** [review]

**Review by BraveMole — CHANGES_REQUESTED** [review]
primo ability needs to be balanced.

Give primo a secondary spit with lesser damage, low spit delay for a slight increase in dps maybe?

**Comment by hyper2snyper** [comment]
> primo ability needs to be balanced.
> 
> Give primo a secondary spit with lesser damage, low spit delay for a slight increase in dps maybe?
> 
> Also, are ai xeno spit working?

i didnt get any pings about ai spit not working. In any case it should be fine considering if you still wanted to spam click spit itd work just the same.

File: code/modules/mob/living/carbon/xenomorph/abilities.dm [hunk 7]
```diff
@@ -771,15 +793,31 @@
...
-	var/turf/current_turf = get_turf(owner)
-
-	if(!current_turf)
+	return
+
+///Starts the xeno firing.
+/datum/action/xeno_action/activable/xeno_spit/proc/start_fire(datum/source, atom/object, turf/location, control, params)
+	SIGNAL_HANDLER
+	var/list/modifiers = params2list(params)
+	if(((modifiers["right"] || modifiers["middle"]) && (modifiers["shift"] || modifiers["ctrl"] || modifiers["left"])) || \
```
[the remark this answers is not in the document]
Comment (BraveMole) [reply code/modules/mob/living/carbon/xenomorph/abilities.dm:802]: Still standing it seems

File: code/modules/mob/living/carbon/xenomorph/castes/spitter/castedatum_spitter.dm [hunk 8]
```diff
@@ -146,4 +146,31 @@
...
+
+	// *** Evolution *** //
+	upgrade_threshold = 1320
+
+	// *** Defense *** //
+	soft_armor = list("melee" = 25, "bullet" = 35, "laser" = 35, "energy" = 35, "bomb" = XENO_BOMB_RESIST_0, "bio" = 20, "rad" = 20, "fire" = 35, "acid" = 20)
+
+	// *** Ranged Attack *** //
+	spit_delay = 0.3 SECONDS
+	spit_types = list(/datum/ammo/xeno/acid/auto)
```
Comment (BraveMole) [inline code/modules/mob/living/carbon/xenomorph/castes/spitter/castedatum_spitter.dm:176]: how about making it so you can choose between old and new spit?

**Review by BraveMole — CHANGES_REQUESTED** [review]

**Comment by vilereaver41** [comment]
please fix the bug that makes it spit like a normal spitter. spitter primo right now makes spitter useless as its spit speed is normal but it keeps the reduction in spit dmg to 10

**Pull request merged**